### PR TITLE
Sockets - Shubha

### PR DESCRIPTION
**Pull request by shubha-rajan**
# JS Adagrams
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
--- | ---
What patterns were you able to use from your Ruby knowledge to apply to JavaScript? | A lot of the logic for the methods in Adagrams was identical to the logic Grace and I used to implement Adagrams in Ruby earlier this cohort. For example, the conditional statements were almost identical. A lot of my work for this assignment was being able to research how to do X in Javascript.
Did you need to use different strategies to find information online about JavaScript than you would for Ruby? | Other than using MDN instead of rubydocs, the process of looking things up was really similar. Only issue was that a lot of StackOverflow answers are really old and propose really roundabout ways of doing things that there are now functions for in ES6.
What was a challenge you were able to overcome on this assignment? | One of the biggest challenges was that my original implementation depended heavily on using hash lookups, and iterating through keys/values simultaneously. After doing some research I found that in JS, maps and objects were the closest thing to hashes. I decided to use maps because they seemed easier to use with the for..of loop structure, and because I wanted to get more experience using them. 
What has been interesting and positive about learning a new programming language? | Learning the differences in ways that certain methods are implemented in different languages. For example, the JS forEach loop, vs Ruby's each loop. 
What is something to focus on your learnings in JavaScript in the next week? |  One thing to focus on for next week is getting used to using higher order functions and callbacks in code instead of writing code sequentially like in Ruby.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -32,6 +32,7 @@
   },
   "dependencies": {
     "core-js": "^2.6.6",
+    "underscore": "^1.9.1",
     "vorpal": "^1.11.4"
```
Comment (droberts-sea) [inline package.json:36]: Good work taking the initiative to use the `underscore` library. For school projects it might be wise to ask first for things like this, just in case struggling with this sort of functionality is part of the point of the project (it wasn't), but in an industry setting this would definitely be the right call.

File: src/adagrams.js [hunk 2]
```diff
@@ -1,8 +1,75 @@
...
+const _ = require('lodash');
+class Adagrams {
+
+
+  static drawLetters() {
+    const letters = [];
+    const letter_freq = new Map([
+      ['A', 9], ['N', 6], ['B', 2], ['O', 8], ['C', 2], ['P', 2], ['D', 4], 
+      ['Q', 1], ['E', 12], ['R', 6], ['F', 2], ['S', 4], ['G', 3], ['T', 6], 
+      ['H', 2], ['U', 4], ['I', 9], ['V', 2], ['J',1], ['W', 2], ['K', 1],
```
Comment (droberts-sea) [inline src/adagrams.js:10]: Why not use an Object here? The syntactic sugar is a little more readable, and I don't see you making heavy use of anything that `Map` gives you.

**Comment by droberts-sea** [comment]
# JS Adagrams
## What We're Looking For



Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | yes
Small commits with meaningful commit messages | yes
**Code Requirements** |
`drawLetters` method | yes
Uses appropriate data structure to store the letter distribution | yes, but see inline
All tests for `drawLetters` pass | yes
`usesAvailableLetters` method | yes
All tests for `usesAvailableLetters` pass | yes
`scoreWord` method | yes
Uses appropriate data structure to store the letter scores | yes
All tests for `scoreWord` pass | yes
`highestScoreFrom` method | yes
Appropriately handles edge cases for tie-breaking logic | yes
All tests for `highestScoreFrom` pass | yes
**Overall** |Great job overall! This code is generally clean and well-organized, and it feels like the JavaScript unit is off to a solid start. Keep up the hard work!

File: src/adagrams.js [hunk 3]
```diff
@@ -1,8 +1,75 @@
...
+    const letters = [];
+    const letter_freq = new Map([
+      ['A', 9], ['N', 6], ['B', 2], ['O', 8], ['C', 2], ['P', 2], ['D', 4], 
+      ['Q', 1], ['E', 12], ['R', 6], ['F', 2], ['S', 4], ['G', 3], ['T', 6], 
+      ['H', 2], ['U', 4], ['I', 9], ['V', 2], ['J',1], ['W', 2], ['K', 1], 
+      ['X', 1], ['L', 4], ['Y',2], ['M', 2], ['Z', 1]]);
+    for (let [key, value] of letter_freq) {
+        for (let i = 0; i < value; i++) {
+              letters.push(key);
+        }
```
Comment (droberts-sea) [inline src/adagrams.js:15]: I really like that you store the letter frequencies and then build the pool on the fly - this is much easier to read than a giant array full of 9 'A's, 2 'B's, etc.
